# Saidak's Proof of the Infinitude of Primes
In 2006, Filip Saidak gave a new, remarkably simple proof that there are an infinite number of primes. You can read it [here](https://fermatslibrary.com/s/a-new-proof-of-euclids-theorem).

It is a constructive proof, which starts with the fact that $N$ and $N+1$ are always coprime, and constructs a sequence $a(n+1) = a(n) \times (a(n) + 1)$ with the property that $a(n)$ has at least $n$ prime factors.

However, the sequence grows so quickly that it is not an efficient way to compute primes.

Let's compute some of the terms in the sequence, which is [A007018](https://oeis.org/A007018) in 
the On-Line Encyclopedia of Integer Sequences.

Start with imports, and a little function to help with formatting.

In [1]:
from sympy import factorint
from tabulate import tabulate

In [2]:
def set_to_str(vals, old_vals):
    s = "{"
    for v in vals:
        if len(s) > 1:
            s += ", "
        if v in old_vals:
            s += f"{v}"
        else:
            s += f"<b>{v}</b>"
    s += "}"
    return s

Now compute $a(n)$ and its prime factors for the first few values of $n$, using `sympy`'s `factorint` function. 

In [3]:
table = []
headers = ["n", "a(n)", "product", "prime factors", "number of distinct prime factors"]
n=0
a=1
prime_factors = list(factorint(a).keys())
table.append([n, a, "", set_to_str(prime_factors, []), len(prime_factors)])
for n in range(1, 8):
    prev_a = a
    prev_prime_factors = prime_factors
    a = a*(a+1)
    prime_factors = list(factorint(a).keys())
    table.append([n, a, f"{prev_a}*{prev_a+1}", set_to_str(prime_factors, prev_prime_factors), len(prime_factors)])
tabulate(table, headers=headers, tablefmt="unsafehtml")

'<table>\n<thead>\n<tr><th style="text-align: right;">  n</th><th style="text-align: right;">                       a(n)</th><th>product                      </th><th>prime factors                                                                     </th><th style="text-align: right;">  number of distinct prime factors</th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;">  0</td><td style="text-align: right;">                          1</td><td>                             </td><td>{}                                                                                </td><td style="text-align: right;">                                 0</td></tr>\n<tr><td style="text-align: right;">  1</td><td style="text-align: right;">                          2</td><td>1*2                          </td><td>{<b>2</b>}                                                                        </td><td style="text-align: right;">                                 1</td></tr>\n<tr><td style="text-align: right;">  2</td><td style="text-align: right;">                          6</td><td>2*3                          </td><td>{2, <b>3</b>}                                                                     </td><td style="text-align: right;">                                 2</td></tr>\n<tr><td style="text-align: right;">  3</td><td style="text-align: right;">                         42</td><td>6*7                          </td><td>{2, 3, <b>7</b>}                                                                  </td><td style="text-align: right;">                                 3</td></tr>\n<tr><td style="text-align: right;">  4</td><td style="text-align: right;">                       1806</td><td>42*43                        </td><td>{2, 3, 7, <b>43</b>}                                                              </td><td style="text-align: right;">                                 4</td></tr>\n<tr><td style="text-align: right;">  5</td><td style="text-align: right;">                    3263442</td><td>1806*1807                    </td><td>{2, 3, 7, <b>13</b>, 43, <b>139</b>}                                              </td><td style="text-align: right;">                                 6</td></tr>\n<tr><td style="text-align: right;">  6</td><td style="text-align: right;">             10650056950806</td><td>3263442*3263443              </td><td>{2, 3, 7, 13, 43, 139, <b>3263443</b>}                                            </td><td style="text-align: right;">                                 7</td></tr>\n<tr><td style="text-align: right;">  7</td><td style="text-align: right;">113423713055421844361000442</td><td>10650056950806*10650056950807</td><td>{2, 3, 7, 13, 43, 139, <b>547</b>, <b>607</b>, <b>1033</b>, <b>31051</b>, 3263443}</td><td style="text-align: right;">                                11</td></tr>\n</tbody>\n</table>'

The new prime factors for each term are shown in bold. The first four terms each add one prime factor, but $a(5)$ adds two (13 and 139), and furthermore, one (13) is smaller than the largest prime factor from the previous term (43). This is to be expected: there was nothing about the construction that says the new primes are going to be increasing in size.

Since we are factorizing what are essentially arbitrary integers, and the number of digits doubles for each term, it's only possible to compute (and factorize) a few more terms before it becomes unfeasible to go any further.

In [4]:
%%time
n=0
a=1
prime_factors = list(factorint(a).keys())
print(n, a, prime_factors)
for n in range(1, 10):
    a = a*(a+1)
    prime_factors = list(factorint(a).keys())
    print(n, a, prime_factors)

0 1 []
1 2 [2]
2 6 [2, 3]
3 42 [2, 3, 7]
4 1806 [2, 3, 7, 43]
5 3263442 [2, 3, 7, 13, 43, 139]
6 10650056950806 [2, 3, 7, 13, 43, 139, 3263443]
7 113423713055421844361000442 [2, 3, 7, 13, 43, 139, 547, 607, 1033, 31051, 3263443]
8 12864938683278671740537145998360961546653259485195806 [2, 3, 7, 13, 43, 139, 547, 607, 1033, 31051, 67003, 29881, 3263443, 9119521, 6212157481]
9 165506647324519964198468195444439180017513152706377497841851388766535868639572406808911988131737645185442 [2, 3, 7, 13, 43, 139, 547, 607, 1033, 31051, 67003, 29881, 3263443, 9119521, 6212157481, 5295435634831, 31401519357481261, 77366930214021991992277]
CPU times: user 1min 42s, sys: 182 ms, total: 1min 42s
Wall time: 1min 43s


### Further reading
* [My Favorite Theorem with Ken Ribet](https://kpknudson.com/my-favorite-theorem/2018/7/10/episode-22-ken-ribet) - talking about different proofs of the infinitude of primes.